Collection of usefull =`R`= scripts; README has to be improved in time

* DONE DEXSeq_plot.R
  CLOSED: [2015-09-02 Wed 20:21]

Script providing extended functionality to plot =`DEXSeq`= results.

** example

#+BEGIN_SRC R :session DEX :exports both :results graphics :file plots/DEXSeq_plot.png :width 800 :height 600
library(DEXSeq)
library(data.table)
source('~/Documents/R_scripts/R/DEXSeq_plot.R')
load( "~/Documents/Cigene/Side_Projects/Radioactive/data/de_exon.RData")

# read gff
gff <- fread('~/Documents/Cigene/Side_Projects/Radioactive/plot-test/Salmon_3p6_Chr_070715_All.filter.gff3', 
             verbose = FALSE)

plot.DEX(dxr, geneID = 'CIGSSA_022765', fitExpToVar = 'group', 
         gff = gff, geneNAME = 'tp53')
#+END_SRC

#+RESULTS:
[[file:plots/DEXSeq_plot.png]]

** TASKS

*** TODO Color option for groups


* DONE fancy_tree
  CLOSED: [2016-01-05 Tue 12:47]

Script to show the alignments side by side with the tree. 
 - *Important* the `plot.tb` accepts accepts additional parameters for
   the `plot.phylo` function (passed on by `...`).


#+BEGIN_SRC R :session DEX :exports both :results graphics :file plots/Fancy_tree.png :width 700 :height 800
library(phangorn)
source('~/Documents/R_scripts/R/fancy_tree.R')

# read alignment
algn <- read.phyDat('data/grp101740_MAFFT.algn', type = 'AA', format = 'fasta')
dm = dist.ml(algn, model = 'WAG')
tree1 <- nj(dm)

# Strange; but necessary to get the correct order
tree1 <- read.tree(text = write.tree(ladderize(tree1)))
# root
tree1 <- root( tree1, outgroup = "Locu.ENSLOCP00000012114")

#grid.newpage()
plot.tb(tree1, algn, block.col = 'steelblue', block.lwd = 1)

#+END_SRC

#+RESULTS:
[[file:plots/Fancy_tree.png]]



* DONE DB_schema.R
  CLOSED: [2015-09-02 Wed 20:21]

Old script with some =`grid`= code to plot DB schemas

* GOfunctions.R

* GOfunctions2.R

* TODO circos_script.R

* TODO circos_script2.R

* grid_heatmap.R

* DONE grid_matrix.R
  CLOSED: [2015-09-02 Wed 20:22]

Plot overlap matrix 

* tile_heatmap.R
